assets/ca/ca.querytranslator.js

```diff
@@ -0,0 +1,387 @@
+     */
+	escapeValue = function (value) {
+		return value.replace(/([\-\+&\|!\(\)\{}\[\]\^"~\*\?:\\])/, '\\$1');
+	};
+
+	/**
+	 * Convert a set of rules from the jQuery query builder into a CA search query.  Performs the inverse operation to
+	 * `convertSearchQueryToQueryBuilderRules`.
+	 * @param {Object} ruleSet
+	 * @returns {String}
+	 */
+	caUI.convertQueryBuilderRuleSetToSearchQuery = function (ruleSet) {
+		if (ruleSet.condition && ruleSet.rules) {
+			return '(' + $.map(ruleSet.rules, caUI.convertQueryBuilderRuleSetToSearchQuery).join(' ' + ruleSet.condition + ' ') + ')';
+		}
+		if (ruleSet.operator && ruleSet.field) {
+			// Escape value to allow special characters
+			var negation = ruleSet.operator.match(/not_/),
+				prefix = ruleSet.field + (negation ? ':-' : ':');
+			switch (negation ? ruleSet.operator.replace('not_', '') : ruleSet.operator) {
+				case 'equal':
+					return prefix + '"' + escapeValue(ruleSet.value) + '"';
+				case 'in':
+					return prefix + '(' + escapeValue(ruleSet.value) + ')';
+				case 'between':
+					return prefix + '[' + escapeValue(ruleSet.value[0]) + ' TO ' + escapeValue(ruleSet.value[1]) + ']';
+				case 'begins_with':
+					return prefix + '"' + escapeValue(ruleSet.value) + '*"';
+				case 'contains':
+					return prefix + '"*' + escapeValue(ruleSet.value) + '*"';
+				case 'ends_with':
+					return prefix + '"*' + escapeValue(ruleSet.value) + '"';
+				case 'is_empty':
+				case 'is_null':
+					return prefix + '*';
+			}
+			return ruleSet.field + ':' + ruleSet.value;
+		}
+		return '';
+	};
+
+	/**
+	 * Retrieve the list of tokens from the given query string.
+	 * @param {String} query
+	 * @return {Array}
+     */
+	getTokenList = function (query) {
+		var token,
+			tokens = [],
+			queryArray = query.trim().split('');
+		while (token = shiftToken(queryArray)) {
+			tokens.push(token);
+		}
+		while (tokens.length >= 2 && tokens[0].type === TOKEN_LPAREN && tokens[tokens.length - 1].type === TOKEN_RPAREN) {
+			tokens = tokens.slice(1, tokens.length - 1);
+		}
+		return tokens;
+	};
+
+	/**
+	 * Retrieve the next token from the `queryArray`, which is destructively processed.
+	 * @param {Array} queryArray
+	 * @returns {Object|undefined}
+	 */
+	shiftToken = function (queryArray) {
+		var character, token,
+			quoted = false,
+			escaped = false,
+			end = false;
+		// End condition.
+		if (queryArray.length === 0) {
+			return undefined;
+		}
+		// Inspect the first character to determine the type of token.
+		character = queryArray.shift();
+		switch (character) {
+			// Single-character tokens.
+			case '(':
+				token = { type: TOKEN_LPAREN };
+				end = true;
+				break;
+			case ')':
+				token = { type: TOKEN_RPAREN };
+				end = true;
+				break;
+			case ':':
+				token = { type: TOKEN_COLON };
+				end = true;
+				break;
+			case '-':
+				token = { type: TOKEN_NEGATION };
+				end = true;
+				break;
+			// Collapse whitespace.
+			case ' ':
+				while (queryArray[0] === ' ') {
```

Comment:
This is one place where having a string would actually be better than having an array, `query = query.replace(/^\s+/, '');`.  However either way, it doesn't change the fact that the `case` is only matching on the space character, so to be technically correct I'd have to add in all the things that match `\s`, which includes stuff you wouldn't probably expect (em space and en space are things iirc).  Can be done but I think only supporting normal space characters is enough.  Otherwise, perhaps at the very start, just do `query = query.replace(/\s+/, ' ');`, then all the spaces will be normal spaces anyway, and also the spaces will be collapsed so this `while` loop will be unnecessary.  

Thoughts?

Do you think the type of whitespace will ever be significant in the search query (i.e. I want to find everything containing a tab character...)?  Does that even work?